pkg/isb/jetstream/writer.go

```diff
@@ -137,37 +137,70 @@ func (jw *jetStreamWriter) Close() error {
 func (jw *jetStreamWriter) Write(_ context.Context, messages []isb.Message) ([]isb.Offset, []error) {
 	labels := map[string]string{"buffer": jw.GetName()}
 	var errs = make([]error, len(messages))
+	for i := 0; i < len(errs); i++ {
+		errs[i] = fmt.Errorf("unknown error")
+	}
 	var writeOffsets = make([]isb.Offset, len(messages))
 	if jw.isFull.Load() {
 		jw.log.Debugw("Is full")
-		isbIsFull.With(map[string]string{"buffer": jw.GetName()}).Inc()
-		for i := range errs {
+		isbFull.With(map[string]string{"buffer": jw.GetName()}).Inc()
+		for i := 0; i < len(errs); i++ {
 			errs[i] = isb.BufferWriteErr{Name: jw.name, Full: true, Message: "Buffer full!"}
 		}
 		isbWriteErrors.With(labels).Inc()
 		return nil, errs
 	}
 
+	var futures = make([]nats.PubAckFuture, len(messages))
+	for index, message := range messages {
+		m := &nats.Msg{
+			Header:  convert2NatsMsgHeader(message.Header),
+			Subject: jw.subject,
+			Data:    message.Payload,
+		}
+		if future, err := jw.js.PublishMsgAsync(m, nats.MsgId(message.Header.ID)); err != nil { // nats.MsgId() is for exactly-once writing
+			errs[index] = err
+		} else {
+			futures[index] = future
+		}
+	}
+	futureCheckDone := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 	wg := new(sync.WaitGroup)
-	for index, msg := range messages {
+	for index, f := range futures {
+		if f == nil {
+			continue
+		}
 		wg.Add(1)
-		go func(message isb.Message, idx int) {
+		go func(idx int, fu nats.PubAckFuture) {
 			defer wg.Done()
-			m := &nats.Msg{
-				Header:  convert2NatsMsgHeader(message.Header),
-				Subject: jw.subject,
-				Data:    message.Payload,
-			}
-			if pubAck, err := jw.js.PublishMsg(m, nats.MsgId(message.Header.ID)); err != nil { // nats.MsgId() is for exactly-once writing
-				errs[idx] = err
-				isbWriteErrors.With(labels).Inc()
-			} else {
+			select {
+			case pubAck := <-fu.Ok():
 				writeOffsets[idx] = &writeOffset{seq: pubAck.Sequence}
+				errs[idx] = nil
 				jw.log.Debugw("Succeeded to publish a message", zap.String("stream", pubAck.Stream), zap.Any("seq", pubAck.Sequence), zap.Bool("duplicate", pubAck.Duplicate), zap.String("domain", pubAck.Domain))
+			case err := <-fu.Err():
+				errs[idx] = err
+				isbWriteErrors.With(labels).Inc()
+			case <-ctx.Done():
```

Comment:
i think we might need to add a timeout here too, what if we lost `fu.Ok()` and/or `fu.Err()`? we will end up waiting indefinitely.. perhaps you can use context with timeout?

Reply:
That is not needed. 

There's a timeout at line 198, which will issue a `cancel()` to end this.

Reply:
you are right.